// payload/dist/admin/components/forms/Form/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useFormModified = exports.useFormProcessing = exports.useFormSubmitted = exports.useWatchForm = exports.useForm = exports.ModifiedContext = exports.ProcessingContext = exports.SubmittedContext = exports.FormWatchContext = exports.FormContext = void 0;
const react_1 = require("react");
const FormContext = (0, react_1.createContext)({});
exports.FormContext = FormContext;
const FormWatchContext = (0, react_1.createContext)({});
exports.FormWatchContext = FormWatchContext;
const SubmittedContext = (0, react_1.createContext)(false);
exports.SubmittedContext = SubmittedContext;
const ProcessingContext = (0, react_1.createContext)(false);
exports.ProcessingContext = ProcessingContext;
const ModifiedContext = (0, react_1.createContext)(false);
exports.ModifiedContext = ModifiedContext;
const useForm = () => (0, react_1.useContext)(FormContext);
exports.useForm = useForm;
const useWatchForm = () => (0, react_1.useContext)(FormWatchContext);
exports.useWatchForm = useWatchForm;
const useFormSubmitted = () => (0, react_1.useContext)(SubmittedContext);
exports.useFormSubmitted = useFormSubmitted;
const useFormProcessing = () => (0, react_1.useContext)(ProcessingContext);
exports.useFormProcessing = useFormProcessing;
const useFormModified = () => (0, react_1.useContext)(ModifiedContext);
exports.useFormModified = useFormModified;
