// payload/dist/collections/operations/local/create.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const getFileByPath_1 = __importDefault(require("../../../uploads/getFileByPath"));
const create_1 = __importDefault(require("../create"));
async function createLocal(payload, options) {
    var _a, _b, _c;
    const { collection: collectionSlug, depth, locale, fallbackLocale, data, user, overrideAccess = true, disableVerificationEmail, showHiddenFields, filePath, file, overwriteExistingFiles = false, req, draft, } = options;
    const collection = payload.collections[collectionSlug];
    return (0, create_1.default)({
        depth,
        data,
        collection,
        overrideAccess,
        disableVerificationEmail,
        showHiddenFields,
        overwriteExistingFiles,
        draft,
        req: {
            ...req || {},
            user,
            payloadAPI: 'local',
            locale: locale || (req === null || req === void 0 ? void 0 : req.locale) || (((_a = payload === null || payload === void 0 ? void 0 : payload.config) === null || _a === void 0 ? void 0 : _a.localization) ? (_c = (_b = payload === null || payload === void 0 ? void 0 : payload.config) === null || _b === void 0 ? void 0 : _b.localization) === null || _c === void 0 ? void 0 : _c.defaultLocale : null),
            fallbackLocale: fallbackLocale || (req === null || req === void 0 ? void 0 : req.fallbackLocale) || null,
            payload,
            files: {
                file: file !== null && file !== void 0 ? file : (0, getFileByPath_1.default)(filePath),
            },
        },
    });
}
exports.default = createLocal;
