// payload/dist/collections/operations/local/findByID.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const findByID_1 = __importDefault(require("../findByID"));
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function findByIDLocal(payload, options) {
    var _a, _b, _c;
    const { collection: collectionSlug, depth, currentDepth, id, locale, fallbackLocale, user, overrideAccess = true, disableErrors = false, showHiddenFields, req, draft = false, } = options;
    const collection = payload.collections[collectionSlug];
    const reqToUse = {
        user: undefined,
        ...req || {},
        payloadAPI: 'local',
        locale: locale || (req === null || req === void 0 ? void 0 : req.locale) || (((_a = payload === null || payload === void 0 ? void 0 : payload.config) === null || _a === void 0 ? void 0 : _a.localization) ? (_c = (_b = payload === null || payload === void 0 ? void 0 : payload.config) === null || _b === void 0 ? void 0 : _b.localization) === null || _c === void 0 ? void 0 : _c.defaultLocale : null),
        fallbackLocale: fallbackLocale || (req === null || req === void 0 ? void 0 : req.fallbackLocale) || null,
        payload,
    };
    if (typeof user !== 'undefined')
        reqToUse.user = user;
    return (0, findByID_1.default)({
        depth,
        currentDepth,
        id,
        collection,
        overrideAccess,
        disableErrors,
        showHiddenFields,
        req: reqToUse,
        draft,
    });
}
exports.default = findByIDLocal;
