// payload/dist/mongoose/buildSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable no-use-before-define */
const mongoose_1 = require("mongoose");
const types_1 = require("../fields/config/types");
const sortableFieldTypes_1 = __importDefault(require("../fields/sortableFieldTypes"));
const flattenTopLevelFields_1 = __importDefault(require("../utilities/flattenTopLevelFields"));
const setBlockDiscriminators = (fields, schema, config, buildSchemaOptions) => {
    fields.forEach((field) => {
        const blockFieldType = field;
        if (blockFieldType.type === 'blocks' && blockFieldType.blocks && blockFieldType.blocks.length > 0) {
            blockFieldType.blocks.forEach((blockItem) => {
                let blockSchemaFields = {};
                blockItem.fields.forEach((blockField) => {
                    const fieldSchema = fieldToSchemaMap[blockField.type];
                    if (fieldSchema) {
                        blockSchemaFields = fieldSchema(blockField, blockSchemaFields, config, buildSchemaOptions);
                    }
                });
                const blockSchema = new mongoose_1.Schema(blockSchemaFields, { _id: false, id: false });
                if (blockFieldType.localized && config.localization) {
                    config.localization.locales.forEach((locale) => {
                        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                        // @ts-ignore Possible incorrect typing in mongoose types, this works
                        schema.path(`${field.name}.${locale}`).discriminator(blockItem.slug, blockSchema);
                        setBlockDiscriminators(blockItem.fields, blockSchema, config, buildSchemaOptions);
                    });
                }
                else {
                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore Possible incorrect typing in mongoose types, this works
                    schema.path(field.name).discriminator(blockItem.slug, blockSchema);
                    setBlockDiscriminators(blockItem.fields, blockSchema, config, buildSchemaOptions);
                }
            });
        }
    });
};
const formatBaseSchema = (field, buildSchemaOptions) => ({
    sparse: field.unique && field.localized,
    unique: (!buildSchemaOptions.disableUnique && field.unique) || false,
    required: false,
    index: field.index || field.unique || false,
});
const localizeSchema = (field, schema, localization) => {
    if (field.localized && localization && Array.isArray(localization.locales)) {
        return {
            type: localization.locales.reduce((localeSchema, locale) => ({
                ...localeSchema,
                [locale]: schema,
            }), {
                _id: false,
            }),
            localized: true,
            index: schema.index,
        };
    }
    return schema;
};
const buildSchema = (config, configFields, buildSchemaOptions = {}) => {
    var _a;
    const { allowIDField, options } = buildSchemaOptions;
    let fields = {};
    let schemaFields = configFields;
    const indexFields = [];
    if (!allowIDField) {
        const idField = schemaFields.find((field) => (0, types_1.fieldAffectsData)(field) && field.name === 'id');
        if (idField) {
            fields = {
                _id: idField.type === 'number' ? Number : String,
            };
            schemaFields = schemaFields.filter((field) => (0, types_1.fieldAffectsData)(field) && field.name !== 'id');
        }
    }
    schemaFields.forEach((field) => {
        if (!(0, types_1.fieldIsPresentationalOnly)(field)) {
            const fieldSchema = fieldToSchemaMap[field.type];
            if (fieldSchema) {
                fields = fieldSchema(field, fields, config, buildSchemaOptions);
            }
            // geospatial field index must be created after the schema is created
            if (fieldIndexMap[field.type]) {
                indexFields.push(...fieldIndexMap[field.type](field, config));
            }
            if (config.indexSortableFields && !buildSchemaOptions.global && !field.index && !field.hidden && sortableFieldTypes_1.default.indexOf(field.type) > -1 && (0, types_1.fieldAffectsData)(field)) {
                indexFields.push({ [field.name]: 1 });
            }
        }
    });
    if ((_a = buildSchemaOptions === null || buildSchemaOptions === void 0 ? void 0 : buildSchemaOptions.options) === null || _a === void 0 ? void 0 : _a.timestamps) {
        indexFields.push({ createdAt: 1 });
        indexFields.push({ updatedAt: 1 });
    }
    const schema = new mongoose_1.Schema(fields, options);
    indexFields.forEach((index) => {
        schema.index(index);
    });
    setBlockDiscriminators((0, flattenTopLevelFields_1.default)(configFields), schema, config, buildSchemaOptions);
    return schema;
};
const fieldIndexMap = {
    point: (field, config) => {
        let index;
        if (field.index === true || field.index === undefined) {
            index = '2dsphere';
        }
        if (field.localized && config.localization) {
            return config.localization.locales.map((locale) => ({ [`${field.name}.${locale}`]: index }));
        }
        return [{ [field.name]: index }];
    },
};
const fieldToSchemaMap = {
    number: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: Number };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    text: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: String };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    email: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: String };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    textarea: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: String };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    richText: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: mongoose_1.Schema.Types.Mixed };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    code: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: String };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    point: (field, fields, config) => {
        const baseSchema = {
            type: {
                type: String,
                enum: ['Point'],
            },
            coordinates: {
                type: [Number],
                sparse: field.unique && field.localized,
                unique: field.unique || false,
                required: false,
                default: field.defaultValue || undefined,
            },
        };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    radio: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = {
            ...formatBaseSchema(field, buildSchemaOptions),
            type: String,
            enum: field.options.map((option) => {
                if (typeof option === 'object')
                    return option.value;
                return option;
            }),
        };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    checkbox: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: Boolean };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    date: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = { ...formatBaseSchema(field, buildSchemaOptions), type: Date };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    upload: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = {
            ...formatBaseSchema(field, buildSchemaOptions),
            type: mongoose_1.Schema.Types.Mixed,
            ref: field.relationTo,
        };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    relationship: (field, fields, config, buildSchemaOptions) => {
        const hasManyRelations = Array.isArray(field.relationTo);
        let schemaToReturn = {};
        if (field.localized && config.localization) {
            schemaToReturn = {
                type: config.localization.locales.reduce((locales, locale) => {
                    let localeSchema = {};
                    if (hasManyRelations) {
                        localeSchema._id = false;
                        localeSchema.value = {
                            type: mongoose_1.Schema.Types.Mixed,
                            refPath: `${field.name}.${locale}.relationTo`,
                        };
                        localeSchema.relationTo = { type: String, enum: field.relationTo };
                    }
                    else {
                        localeSchema = {
                            ...formatBaseSchema(field, buildSchemaOptions),
                            type: mongoose_1.Schema.Types.Mixed,
                            ref: field.relationTo,
                        };
                    }
                    return {
                        ...locales,
                        [locale]: field.hasMany ? [localeSchema] : localeSchema,
                    };
                }, {}),
                localized: true,
            };
        }
        else if (hasManyRelations) {
            schemaToReturn._id = false;
            schemaToReturn.value = {
                type: mongoose_1.Schema.Types.Mixed,
                refPath: `${field.name}.relationTo`,
            };
            schemaToReturn.relationTo = { type: String, enum: field.relationTo };
            if (field.hasMany)
                schemaToReturn = [schemaToReturn];
        }
        else {
            schemaToReturn = {
                ...formatBaseSchema(field, buildSchemaOptions),
                type: mongoose_1.Schema.Types.Mixed,
                ref: field.relationTo,
            };
            if (field.hasMany)
                schemaToReturn = [schemaToReturn];
        }
        return {
            ...fields,
            [field.name]: schemaToReturn,
        };
    },
    row: (field, fields, config, buildSchemaOptions) => {
        let newFields = { ...fields };
        field.fields.forEach((subField) => {
            const fieldSchema = fieldToSchemaMap[subField.type];
            if (fieldSchema) {
                newFields = fieldSchema(subField, newFields, config, buildSchemaOptions);
            }
        });
        return newFields;
    },
    collapsible: (field, fields, config, buildSchemaOptions) => {
        let newFields = { ...fields };
        field.fields.forEach((subField) => {
            const fieldSchema = fieldToSchemaMap[subField.type];
            if (fieldSchema) {
                newFields = fieldSchema(subField, newFields, config, buildSchemaOptions);
            }
        });
        return newFields;
    },
    tabs: (field, fields, config, buildSchemaOptions) => {
        let newFields = { ...fields };
        field.tabs.forEach((tab) => {
            tab.fields.forEach((subField) => {
                const fieldSchema = fieldToSchemaMap[subField.type];
                if (fieldSchema) {
                    newFields = fieldSchema(subField, newFields, config, buildSchemaOptions);
                }
            });
        });
        return newFields;
    },
    array: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = {
            ...formatBaseSchema(field, buildSchemaOptions),
            type: [buildSchema(config, field.fields, {
                    options: { _id: false, id: false },
                    allowIDField: true,
                    disableUnique: buildSchemaOptions.disableUnique,
                })],
        };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    group: (field, fields, config, buildSchemaOptions) => {
        var _a, _b;
        let { required } = field;
        if (((_a = field === null || field === void 0 ? void 0 : field.admin) === null || _a === void 0 ? void 0 : _a.condition) || (field === null || field === void 0 ? void 0 : field.localized) || ((_b = field === null || field === void 0 ? void 0 : field.access) === null || _b === void 0 ? void 0 : _b.create))
            required = false;
        const formattedBaseSchema = formatBaseSchema(field, buildSchemaOptions);
        const baseSchema = {
            ...formattedBaseSchema,
            required: required && field.fields.some((subField) => { var _a, _b; return (!(0, types_1.fieldIsPresentationalOnly)(subField) && subField.required && !subField.localized && !((_a = subField === null || subField === void 0 ? void 0 : subField.admin) === null || _a === void 0 ? void 0 : _a.condition) && !((_b = subField === null || subField === void 0 ? void 0 : subField.access) === null || _b === void 0 ? void 0 : _b.create)); }),
            type: buildSchema(config, field.fields, {
                options: {
                    _id: false,
                    id: false,
                },
                disableUnique: buildSchemaOptions.disableUnique,
            }),
        };
        return {
            ...fields,
            [field.name]: localizeSchema(field, baseSchema, config.localization),
        };
    },
    select: (field, fields, config, buildSchemaOptions) => {
        const baseSchema = {
            ...formatBaseSchema(field, buildSchemaOptions),
            type: String,
            enum: field.options.map((option) => {
                if (typeof option === 'object')
                    return option.value;
                return option;
            }),
        };
        const schemaToReturn = localizeSchema(field, baseSchema, config.localization);
        return {
            ...fields,
            [field.name]: field.hasMany ? [schemaToReturn] : schemaToReturn,
        };
    },
    blocks: (field, fields, config) => {
        const baseSchema = [new mongoose_1.Schema({}, { _id: false, discriminatorKey: 'blockType' })];
        let schemaToReturn;
        if (field.localized && config.localization) {
            schemaToReturn = config.localization.locales.reduce((localeSchema, locale) => ({
                ...localeSchema,
                [locale]: baseSchema,
            }), {});
        }
        else {
            schemaToReturn = baseSchema;
        }
        return {
            ...fields,
            [field.name]: schemaToReturn,
        };
    },
};
exports.default = buildSchema;
